File: python/src/aoc/year2015/day10.py
from aoc.util import load_input


def step(seq):
    """
    >>> step('1')
    '11'
    >>> step('11')
    '21'
    >>> step('21')
    '1211'
    >>> step('1211')
    '111221'
    >>> step('111221')
    '312211'
    """
    result = ''
    count = 0
    c = lc = seq[0]
    for c in seq:
        if c == lc:
            count += 1
        else:
            result += str(count) + lc
            count = 1
            lc = c
    result += str(count) + c
    return result


def play_look_and_say(start_seq, rounds):
    seq = start_seq
    for i in range(rounds):
        seq = step(seq)
    return len(seq)


def part1(lines):
    start_seq = lines[0]
    return play_look_and_say(start_seq, 40)


def part2(lines):
    start_seq = lines[0]
    return play_look_and_say(start_seq, 50)


if __name__ == "__main__":
    data = load_input(__file__, 2015, '10')
    print(part1(data))
    print(part2(data))

File: kotlin/src/main/kotlin/io/github/ocirne/aoc/year2015/Day10.kt
package io.github.ocirne.aoc.year2015

import io.github.ocirne.aoc.AocChallenge

class Day10(val lines: List<String>) : AocChallenge(2015, 10) {

    override fun part1(): Int {
        return playLookAndSay(40)
    }

    override fun part2(): Int {
        return playLookAndSay(50)
    }

    private fun playLookAndSay(rounds: Int): Int {
        var seq = lines.first().toList()
        repeat(rounds) { seq = step(seq) }
        return seq.size
    }

    fun step(seq: List<Char>): List<Char> {
        val result = ArrayList<Char>()
        var count = 0
        var lc = seq.first()
        seq.forEach { c ->
            if (c == lc) {
                count++
            } else {
                add(result, count, lc)
                count = 1
                lc = c
            }
        }
        add(result, count, lc)
        return result
    }

    private fun add(result: ArrayList<Char>, count: Int, lastCharacter: Char) {
        result.addAll(count.toString().toList())
        result.add(lastCharacter)
    }
}